--- a/components/esp_hal_twai/include/hal/twai_hal.h
+++ b/components/esp_hal_twai/include/hal/twai_hal.h
@@ -280,9 +280,7 @@
 __attribute__((always_inline))
 static inline twai_error_flags_t twai_hal_get_err_flags(twai_hal_context_t *hal_ctx)
 {
-    twai_error_flags_t error_flags = hal_ctx->errors;
-    hal_ctx->errors.val = 0;    // clear ctx for next errors
-    return error_flags;
+    return hal_ctx->errors;
 }
 
 /* ------------------------------- TX and RX -------------------------------- */
--- a/components/esp_hal_twai/twai_hal_v2.c
+++ b/components/esp_hal_twai/twai_hal_v2.c
@@ -4,7 +4,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#include <sys/param.h>
+#include <zephyr/sys/util.h>
 #include "hal/twai_hal.h"
 #include "hal/twaifd_ll.h"
 
