--- a/components/esp_hal_dma/dw_gdma_hal.c
+++ b/components/esp_hal_dma/dw_gdma_hal.c
@@ -5,7 +5,7 @@
  */
 
 #include <stdlib.h>
-#include <sys/param.h>
+#include <zephyr/sys/util.h>
 #include "hal/dw_gdma_hal.h"
 #include "hal/dw_gdma_ll.h"
 
--- a/components/esp_hal_dma/esp32p4/include/hal/gdma_ll.h
+++ b/components/esp_hal_dma/esp32p4/include/hal/gdma_ll.h
@@ -12,7 +12,7 @@
 
 #include <stdint.h>
 #include <stdbool.h>
-#include <sys/param.h>
+#include <zephyr/sys/util.h>
 #include "soc/hp_sys_clkrst_struct.h"
 #include "soc/soc_etm_source.h"
 #include "hal/config.h"
--- a/components/esp_hal_dma/esp32s31/include/hal/gdma_ll.h
+++ b/components/esp_hal_dma/esp32s31/include/hal/gdma_ll.h
@@ -7,7 +7,7 @@
 
 #include <stdint.h>
 #include <stdbool.h>
-#include <sys/param.h>
+#include <zephyr/sys/util.h>
 #include "soc/hp_sys_clkrst_struct.h"
 #include "soc/lp_peri_clkrst_struct.h"
 #include "soc/soc_etm_source.h"
