--- a/components/esp_rom/esp32c2/include/esp32c2/rom/uart.h
+++ b/components/esp_rom/esp32c2/include/esp32c2/rom/uart.h
@@ -113,6 +113,10 @@
     XON_XOFF_CTRL
 } UartFlowCtrl;
 
+/* Zephyr defines EMPTY as a macro in util_macro.h, so we need to undefine it here */
+#ifdef EMPTY
+#undef EMPTY
+#endif
 typedef enum {
     EMPTY,
     UNDER_WRITE,
--- a/components/esp_rom/esp32c2/ld/esp32c2.rom.libc.ld
+++ b/components/esp_rom/esp32c2/ld/esp32c2.rom.libc.ld
@@ -31,8 +31,12 @@
 strsep = 0x40000540;
 strspn = 0x40000544;
 strtok_r = 0x40000548;
-longjmp = 0x40000550;
-setjmp = 0x40000554;
+
+/* ZEPHYR: Keep PROVIDE for these symbols: */
+PROVIDE ( longjmp = 0x40000550 );
+PROVIDE ( setjmp = 0x40000554 );
+/*******************************************/
+
 abs = 0x40000558;
 div = 0x4000055c;
 labs = 0x40000560;
--- a/components/esp_rom/esp32c2/ld/esp32c2.rom.newlib.ld
+++ b/components/esp_rom/esp32c2/ld/esp32c2.rom.newlib.ld
@@ -20,49 +20,29 @@
  Group newlib
  ***************************************/
 
+/* ZEPHYR: Keep PROVIDE for these symbols: */
+PROVIDE ( strdup = 0x40000510 );
+PROVIDE ( strndup = 0x40000534 );
+PROVIDE ( rand_r = 0x4000056c );
+PROVIDE ( rand = 0x40000570 );
+PROVIDE ( srand = 0x40000574 );
+PROVIDE ( atoi = 0x40000580 );
+PROVIDE ( atol = 0x40000584 );
+PROVIDE ( strtol = 0x40000588 );
+PROVIDE ( strtoul = 0x4000058c );
+PROVIDE ( fflush = 0x40000590 );
+PROVIDE ( _fflush_r = 0x40000594 );
+PROVIDE ( _fwalk = 0x40000598 );
+PROVIDE ( _fwalk_reent = 0x4000059c );
+PROVIDE ( __swbuf_r = 0x400005a8 );
+/*******************************************/
+
 /* Functions */
 _isatty_r = 0x400004b4;
-strdup = 0x40000510;
-strndup = 0x40000534;
-rand_r = 0x4000056c;
-rand = 0x40000570;
-srand = 0x40000574;
-atoi = 0x40000580;
-atol = 0x40000584;
-strtol = 0x40000588;
-strtoul = 0x4000058c;
-fflush = 0x40000590;
-_fflush_r = 0x40000594;
-_fwalk = 0x40000598;
-_fwalk_reent = 0x4000059c;
 __smakebuf_r = 0x400005a0;
 __swhatbuf_r = 0x400005a4;
-__swbuf_r = 0x400005a8;
 __swbuf = 0x400005ac;
 __swsetup_r = 0x400005b0;
-_strtod_l = 0x400005b4;
-_strtod_r = 0x400005b8;
-strtod_l = 0x400005bc;
-strtod = 0x400005c0;
-strtof_l = 0x400005c4;
-strtof = 0x400005c8;
-_strtol_r = 0x400005cc;
-strtol_l = 0x400005d0;
-_strtoul_r = 0x400005d4;
-strtoul_l = 0x400005d8;
-asctime = 0x40000638;
-asctime_r = 0x4000063c;
-atof = 0x40000640;
-atoff = 0x40000644;
-_dtoa_r = 0x40000648;
-_wctomb_r = 0x4000064c;
-__ascii_wctomb = 0x40000650;
-_mbtowc_r = 0x40000654;
-__ascii_mbtowc = 0x40000658;
-puts = 0x4000065c;
-putc = 0x40000660;
-putchar = 0x40000664;
-__errno = 0x40000670;
 toupper = 0x400004ec;
 tolower = 0x400004f0;
 isalnum = 0x400004bc;
--- a/components/esp_rom/esp32c3/include/esp32c3/rom/uart.h
+++ b/components/esp_rom/esp32c3/include/esp32c3/rom/uart.h
@@ -113,6 +113,11 @@
     XON_XOFF_CTRL
 } UartFlowCtrl;
 
+/* Undef EMPTY if defined by Zephyr util_macro.h */
+#ifdef EMPTY
+#undef EMPTY
+#endif
+
 typedef enum {
     EMPTY,
     UNDER_WRITE,
--- a/components/esp_rom/esp32c3/ld/esp32c3.rom.libc-suboptimal_for_misaligned_mem.ld
+++ b/components/esp_rom/esp32c3/ld/esp32c3.rom.libc-suboptimal_for_misaligned_mem.ld
@@ -7,6 +7,7 @@
  * See details in the https://sourceware.org/pipermail/newlib/2025/021470.html
  * patch series.  */
 memcpy = 0x40000358;
+memccpy = 0x400003c4;
 memmove = 0x4000035c;
 memcmp = 0x40000360;
 strcpy = 0x40000364;
--- a/components/esp_rom/esp32c3/ld/esp32c3.rom.newlib.ld
+++ b/components/esp_rom/esp32c3/ld/esp32c3.rom.newlib.ld
@@ -19,21 +19,24 @@
  Group newlib
  ***************************************/
 
-/* Functions */
-strdup = 0x400003dc;
-strndup = 0x40000400;
-rand_r = 0x40000438;
-rand = 0x4000043c;
-srand = 0x40000440;
-atoi = 0x4000044c;
-atol = 0x40000450;
-strtol = 0x40000454;
-strtoul = 0x40000458;
+/* ZEPHYR: Keep PROVIDE for these symbols: */
+PROVIDE ( strdup = 0x400003dc );
+PROVIDE ( strndup = 0x40000400 );
+PROVIDE ( rand = 0x4000043c );
+PROVIDE ( srand = 0x40000440 );
+PROVIDE ( rand_r = 0x40000438 );
+PROVIDE ( atoi = 0x4000044c );
+PROVIDE ( atol = 0x40000450 );
+PROVIDE ( strtol = 0x40000454 );
+PROVIDE ( strtoul = 0x40000458 );
 PROVIDE( fflush = 0x4000045c );
 PROVIDE( _fflush_r = 0x40000460 );
 PROVIDE( _fwalk = 0x40000464 );
 PROVIDE( _fwalk_reent = 0x40000468 );
 PROVIDE( __swbuf_r = 0x40000474 );
+/*******************************************/
+
+/* Functions */
 __swbuf = 0x40000478;
 toupper = 0x400003b8;
 tolower = 0x400003bc;
--- a/components/esp_rom/esp32c5/include/esp32c5/rom/uart.h
+++ b/components/esp_rom/esp32c5/include/esp32c5/rom/uart.h
@@ -113,6 +113,10 @@
     XON_XOFF_CTRL
 } UartFlowCtrl;
 
+/* Undef EMPTY if defined by Zephyr util_macro.h */
+#ifdef EMPTY
+#undef EMPTY
+#endif
 typedef enum {
     EMPTY,
     UNDER_WRITE,
--- a/components/esp_rom/esp32c5/ld/esp32c5.rom.eco3.ld
+++ b/components/esp_rom/esp32c5/ld/esp32c5.rom.eco3.ld
@@ -258,10 +258,10 @@
 phy_set_lb_txiq = 0x40001414;
 phy_rxiq_opt = 0x40001418;
 phy_bb_gain_index = 0x40001430;
-//phy_rfrx_gain_index = 0x40001434;
+phy_rfrx_gain_index = 0x40001434;
 phy_gen_rx_gain_table = 0x40001438;
 phy_get_rxbb_dc = 0x4000143c;
-//phy_wr_rx_gain_mem = 0x40001440;
+phy_wr_rx_gain_mem = 0x40001440;
 phy_set_tsens_power_ = 0x40001444;
 phy_get_tsens_value_ = 0x40001448;
 phy_tsens_read_init = 0x4000144c;
--- a/components/esp_rom/esp32c5/ld/esp32c5.rom.libc.ld
+++ b/components/esp_rom/esp32c5/ld/esp32c5.rom.libc.ld
@@ -17,6 +17,11 @@
  Group libc
  ***************************************/
 
+/* ZEPHYR: Keep PROVIDE for these symbols: */
+PROVIDE ( longjmp = 0x40000580 );
+PROVIDE ( setjmp = 0x40000584 );
+/*******************************************/
+
 /* Functions */
 esp_rom_newlib_init_common_mutexes = 0x400004b4;
 memset = 0x400004b8;
@@ -46,8 +51,6 @@
 strsep = 0x40000570;
 strspn = 0x40000574;
 strtok_r = 0x40000578;
-longjmp = 0x40000580;
-setjmp = 0x40000584;
 abs = 0x40000588;
 div = 0x4000058c;
 labs = 0x40000590;
--- a/components/esp_rom/esp32c5/ld/esp32c5.rom.newlib.ld
+++ b/components/esp_rom/esp32c5/ld/esp32c5.rom.newlib.ld
@@ -17,24 +17,27 @@
  Group newlib
  ***************************************/
 
+/* ZEPHYR: Keep PROVIDE for these symbols: */
+PROVIDE ( strdup = 0x40000540 );
+PROVIDE ( strndup = 0x40000564 );
+PROVIDE ( rand = 0x400005a0 );
+PROVIDE ( srand = 0x400005a4 );
+PROVIDE ( rand_r = 0x4000059c );
+PROVIDE ( atoi = 0x400005b0 );
+PROVIDE ( atol = 0x400005b4 );
+PROVIDE ( strtol = 0x400005b8 );
+PROVIDE ( strtoul = 0x400005bc );
+PROVIDE ( fflush = 0x400005c0 );
+PROVIDE ( _fflush_r = 0x400005c4 );
+PROVIDE ( _fwalk = 0x400005c8 );
+PROVIDE ( _fwalk_reent = 0x400005cc );
+PROVIDE ( __swbuf_r = 0x400005d8 );
+/*******************************************/
+
 /* Functions */
 _isatty_r = 0x400004e4;
-strdup = 0x40000540;
-strndup = 0x40000564;
-rand_r = 0x4000059c;
-rand = 0x400005a0;
-srand = 0x400005a4;
-atoi = 0x400005b0;
-atol = 0x400005b4;
-strtol = 0x400005b8;
-strtoul = 0x400005bc;
-fflush = 0x400005c0;
-_fflush_r = 0x400005c4;
-_fwalk = 0x400005c8;
-_fwalk_reent = 0x400005cc;
 __smakebuf_r = 0x400005d0;
 __swhatbuf_r = 0x400005d4;
-__swbuf_r = 0x400005d8;
 __swbuf = 0x400005dc;
 __swsetup_r = 0x400005e0;
 toupper = 0x4000051c;
--- a/components/esp_rom/esp32c61/ld/esp32c61.rom.eco4.ld
+++ b/components/esp_rom/esp32c61/ld/esp32c61.rom.eco4.ld
@@ -143,12 +143,12 @@
 phy_set_freq = 0x400011c8;
 phy_write_pll_cap = 0x400011cc;
 phy_read_pll_cap = 0x400011d0;
-//phy_chip_set_chan_misc = 0x400011d4;
+phy_chip_set_chan_misc = 0x400011d4;
 phy_freq_set_reg = 0x400011d8;
 phy_rfpll_chgp_cal = 0x400011dc;
 phy_gen_rx_gain_table = 0x400011e0;
 phy_get_rxbb_dc = 0x400011e4;
-//phy_wr_rx_gain_mem = 0x400011e8;
+phy_wr_rx_gain_mem = 0x400011e8;
 phy_rfpll_cap_track = 0x400011ec;
 phy_txpwr_correct = 0x400011f4;
 phy_txpwr_cal_track = 0x400011f8;
@@ -288,7 +288,7 @@
 phy_spur_cal = 0x40001434;
 phy_spur_coef_cfg = 0x40001438;
 phy_bb_gain_index = 0x4000143c;
-//phy_rfrx_gain_index = 0x40001440;
+phy_rfrx_gain_index = 0x40001440;
 phy_set_tsens_power_ = 0x40001444;
 phy_set_tsens_range_ = 0x40001448;
 phy_get_tsens_value_ = 0x4000144c;
--- a/components/esp_rom/esp32c6/include/esp32c6/rom/uart.h
+++ b/components/esp_rom/esp32c6/include/esp32c6/rom/uart.h
@@ -113,6 +113,10 @@
     XON_XOFF_CTRL
 } UartFlowCtrl;
 
+/* Undef EMPTY if defined by Zephyr util_macro.h */
+#ifdef EMPTY
+#undef EMPTY
+#endif
 typedef enum {
     EMPTY,
     UNDER_WRITE,
--- a/components/esp_rom/esp32c6/ld/esp32c6.rom.libc.ld
+++ b/components/esp_rom/esp32c6/ld/esp32c6.rom.libc.ld
@@ -3,6 +3,12 @@
  *
  * SPDX-License-Identifier: Apache-2.0
  */
+
+/* ZEPHYR: Keep PROVIDE for these symbols: */
+PROVIDE ( longjmp = 0x40000570 );
+PROVIDE ( setjmp = 0x40000574 );
+/*******************************************/
+
 esp_rom_newlib_init_common_mutexes = 0x400004a4;
 memset = 0x400004a8;
 strlen = 0x400004c8;
@@ -31,8 +37,6 @@
 strsep = 0x40000560;
 strspn = 0x40000564;
 strtok_r = 0x40000568;
-longjmp = 0x40000570;
-setjmp = 0x40000574;
 abs = 0x40000578;
 div = 0x4000057c;
 labs = 0x40000580;
--- a/components/esp_rom/esp32c6/ld/esp32c6.rom.newlib.ld
+++ b/components/esp_rom/esp32c6/ld/esp32c6.rom.newlib.ld
@@ -19,24 +19,27 @@
  Group newlib
  ***************************************/
 
+/* ZEPHYR: Keep PROVIDE for these symbols: */
+PROVIDE ( strdup = 0x40000530 );
+PROVIDE ( strndup = 0x40000554 );
+PROVIDE ( rand = 0x40000590 );
+PROVIDE ( srand = 0x40000594 );
+PROVIDE ( rand_r = 0x4000058c );
+PROVIDE ( atoi = 0x400005a0 );
+PROVIDE ( atol = 0x400005a4 );
+PROVIDE ( strtol = 0x400005a8 );
+PROVIDE ( strtoul = 0x400005ac );
+PROVIDE ( fflush = 0x400005b0 );
+PROVIDE ( _fflush_r = 0x400005b4 );
+PROVIDE ( _fwalk = 0x400005b8 );
+PROVIDE ( _fwalk_reent = 0x400005bc );
+PROVIDE ( __swbuf_r = 0x400005c8 );
+/*******************************************/
+
 /* Functions */
 _isatty_r = 0x400004d4;
-strdup = 0x40000530;
-strndup = 0x40000554;
-rand_r = 0x4000058c;
-rand = 0x40000590;
-srand = 0x40000594;
-atoi = 0x400005a0;
-atol = 0x400005a4;
-strtol = 0x400005a8;
-strtoul = 0x400005ac;
-fflush = 0x400005b0;
-_fflush_r = 0x400005b4;
-_fwalk = 0x400005b8;
-_fwalk_reent = 0x400005bc;
 __smakebuf_r = 0x400005c0;
 __swhatbuf_r = 0x400005c4;
-__swbuf_r = 0x400005c8;
 __swbuf = 0x400005cc;
 __swsetup_r = 0x400005d0;
 toupper = 0x4000050c;
--- a/components/esp_rom/esp32h2/include/esp32h2/rom/uart.h
+++ b/components/esp_rom/esp32h2/include/esp32h2/rom/uart.h
@@ -113,6 +113,10 @@
     XON_XOFF_CTRL
 } UartFlowCtrl;
 
+/* Undef EMPTY if defined by Zephyr util_macro.h */
+#ifdef EMPTY
+#undef EMPTY
+#endif
 typedef enum {
     EMPTY,
     UNDER_WRITE,
--- a/components/esp_rom/esp32h2/ld/esp32h2.rom.libc.ld
+++ b/components/esp_rom/esp32h2/ld/esp32h2.rom.libc.ld
@@ -31,8 +31,9 @@
 strsep = 0x40000558;
 strspn = 0x4000055c;
 strtok_r = 0x40000560;
-longjmp = 0x40000568;
-setjmp = 0x4000056c;
+/* ZEPHYR: Keep PROVIDE for these symbols: */
+PROVIDE ( longjmp = 0x40000568 );
+PROVIDE ( setjmp = 0x4000056c );
 abs = 0x40000570;
 div = 0x40000574;
 labs = 0x40000578;
--- a/components/esp_rom/esp32h2/ld/esp32h2.rom.newlib.ld
+++ b/components/esp_rom/esp32h2/ld/esp32h2.rom.newlib.ld
@@ -21,22 +21,23 @@
 
 /* Functions */
 _isatty_r = 0x400004cc;
-strdup = 0x40000528;
-strndup = 0x4000054c;
-rand_r = 0x40000584;
-rand = 0x40000588;
-srand = 0x4000058c;
-atoi = 0x40000598;
-atol = 0x4000059c;
-strtol = 0x400005a0;
-strtoul = 0x400005a4;
-fflush = 0x400005a8;
-_fflush_r = 0x400005ac;
-_fwalk = 0x400005b0;
-_fwalk_reent = 0x400005b4;
+/* ZEPHYR: Keep PROVIDE for these symbols: */
+PROVIDE ( strdup = 0x40000528 );
+PROVIDE ( strndup = 0x4000054c );
+PROVIDE ( rand_r = 0x40000584 );
+PROVIDE ( rand = 0x40000588 );
+PROVIDE ( srand = 0x4000058c );
+PROVIDE ( atoi = 0x40000598 );
+PROVIDE ( atol = 0x4000059c );
+PROVIDE ( strtol = 0x400005a0 );
+PROVIDE ( strtoul = 0x400005a4 );
+PROVIDE ( fflush = 0x400005a8 );
+PROVIDE ( _fflush_r = 0x400005ac );
+PROVIDE ( _fwalk = 0x400005b0 );
+PROVIDE ( _fwalk_reent = 0x400005b4 );
 __smakebuf_r = 0x400005b8;
 __swhatbuf_r = 0x400005bc;
-__swbuf_r = 0x400005c0;
+PROVIDE ( __swbuf_r = 0x400005c0 );
 __swbuf = 0x400005c4;
 __swsetup_r = 0x400005c8;
 toupper = 0x40000504;
--- a/components/esp_rom/esp32/include/esp32/rom/uart.h
+++ b/components/esp_rom/esp32/include/esp32/rom/uart.h
@@ -111,6 +111,8 @@
     XON_XOFF_CTRL
 } UartFlowCtrl;
 
+/* EMPTY conflicts with Zephyr util_macro.h */
+#undef EMPTY
 typedef enum {
     EMPTY,
     UNDER_WRITE,
--- a/components/esp_rom/esp32s2/include/esp32s2/rom/uart.h
+++ b/components/esp_rom/esp32s2/include/esp32s2/rom/uart.h
@@ -113,6 +113,8 @@
     XON_XOFF_CTRL
 } UartFlowCtrl;
 
+/* Workaround: EMPTY conflicts with Zephyr's util_macro.h */
+#undef EMPTY
 typedef enum {
     EMPTY,
     UNDER_WRITE,
--- a/components/esp_rom/esp32s2/ld/esp32s2.rom.libc-funcs.ld
+++ b/components/esp_rom/esp32s2/ld/esp32s2.rom.libc-funcs.ld
@@ -25,7 +25,8 @@
 ispunct = 0x40007994;
 labs = 0x40000648;
 ldiv = 0x40000650;
-longjmp = 0x400005a4;
+PROVIDE ( longjmp = 0x400005a4 );
+PROVIDE ( setjmp = 0x40000540 );
 memccpy = 0x4001ab00;
 memchr = 0x4001ab24;
 memcmp = 0x4001ab40;
@@ -34,7 +35,6 @@
 memrchr = 0x4001acec;
 memset = 0x4001ad3c;
 qsort = 0x400006f4;
-setjmp = 0x40000540;
 strcat = 0x4001ad90;
 strchr = 0x4001adb0;
 strcmp = 0x40007be4;
@@ -51,6 +51,6 @@
 strsep = 0x4000806c;
 strspn = 0x4001aebc;
 strstr = 0x4001aee8;
-__strtok_r = 0x4001af18;
+PROVIDE ( __strtok_r = 0x4001af18 );
 strtok_r = 0x4001af7c;
 toascii = 0x4001af90;
--- a/components/esp_rom/esp32s2/ld/esp32s2.rom.newlib-reent-funcs.ld
+++ b/components/esp_rom/esp32s2/ld/esp32s2.rom.newlib-reent-funcs.ld
@@ -3,51 +3,51 @@
  *
  * SPDX-License-Identifier: Apache-2.0
  */
-_cleanup_r = 0x4001a480;
+PROVIDE ( _cleanup_r = 0x4001a480 );
 creat = 0x4000788c;
 open = 0x400080c4;
 fclose = 0x4001a804;
-_fclose_r = 0x4001a714;
-fflush = 0x40001bb8;
-_fflush_r = 0x40001b30;
-__fp_unlock_all = 0x4001a64c;
-__fputwc = 0x40001770;
+PROVIDE ( _fclose_r = 0x4001a714 );
+PROVIDE ( fflush = 0x40001bb8 );
+PROVIDE ( _fflush_r = 0x40001b30 );
+PROVIDE ( __fp_unlock_all = 0x4001a64c );
+PROVIDE ( __fputwc = 0x40001770 );
 fputwc = 0x40001864;
-_fputwc_r = 0x400017f8;
-_fwalk = 0x4001bcec;
-_fwalk_reent = 0x4001bd24;
-__sflush_r = 0x400019dc;
-__swbuf = 0x4000167c;
-__swbuf_r = 0x400015bc;
-__swrite = 0x4001a698;
-__sread = 0x4001a660;
-__sseek = 0x4001a6cc;
-srand = 0x40007a24;
-__ascii_mbtowc = 0x40007a04;
-__ascii_wctomb = 0x400018d0;
-_mbtowc_r = 0x400079e0;
-rand_r = 0x40007af4;
-__sclose = 0x4001a700;
-__seofread = 0x4001a690;
+PROVIDE ( _fputwc_r = 0x400017f8 );
+PROVIDE ( _fwalk = 0x4001bcec );
+PROVIDE ( _fwalk_reent = 0x4001bd24 );
+PROVIDE ( __sflush_r = 0x400019dc );
+PROVIDE ( __swbuf = 0x4000167c );
+PROVIDE ( __swbuf_r = 0x400015bc );
+PROVIDE ( __swrite = 0x4001a698 );
+PROVIDE ( __sread = 0x4001a660 );
+PROVIDE ( __sseek = 0x4001a6cc );
+PROVIDE ( srand = 0x40007a24 );
+PROVIDE ( rand_r = 0x40007af4 );
+PROVIDE ( __ascii_mbtowc = 0x40007a04 );
+PROVIDE ( __ascii_wctomb = 0x400018d0 );
+PROVIDE ( _mbtowc_r = 0x400079e0 );
+PROVIDE ( __sclose = 0x4001a700 );
+PROVIDE ( __seofread = 0x4001a690 );
 setlocale = 0x40001c44;
-_setlocale_r = 0x40001bdc;
-__sfmoreglue = 0x4001a4c8;
-__sfp = 0x4001a590;
-__sfp_lock_acquire = 0x4001a508;
-__sfp_lock_release = 0x4001a514;
-__sinit = 0x4001a538;
-__sinit_lock_acquire = 0x4001a520;
-__sinit_lock_release = 0x4001a52c;
-strndup = 0x40007fe8;
-_strndup_r = 0x40007ffc;
+PROVIDE ( _setlocale_r = 0x40001bdc );
+PROVIDE ( __sfmoreglue = 0x4001a4c8 );
+PROVIDE ( __sfp = 0x4001a590 );
+PROVIDE ( __sfp_lock_acquire = 0x4001a508 );
+PROVIDE ( __sfp_lock_release = 0x4001a514 );
+PROVIDE ( __sinit = 0x4001a538 );
+PROVIDE ( __sinit_lock_acquire = 0x4001a520 );
+PROVIDE ( __sinit_lock_release = 0x4001a52c );
+PROVIDE ( strndup = 0x40007fe8 );
+PROVIDE ( strdup = 0x40007d84 );
+PROVIDE ( _strndup_r = 0x40007ffc );
 wcrtomb = 0x400012f4;
-_wcrtomb_r = 0x400012a0;
-_wctomb_r = 0x400018ac;
-strdup = 0x40007d84;
-_strdup_r = 0x40007d98;
-__locale_ctype_ptr = 0x40001c2c;
-__locale_ctype_ptr_l = 0x40001c24;
-__locale_mb_cur_max = 0x40001c0c;
+PROVIDE ( _wcrtomb_r = 0x400012a0 );
+PROVIDE ( _wctomb_r = 0x400018ac );
+PROVIDE ( _strdup_r = 0x40007d98 );
+PROVIDE ( __locale_ctype_ptr = 0x40001c2c );
+PROVIDE ( __locale_ctype_ptr_l = 0x40001c24 );
+PROVIDE ( __locale_mb_cur_max = 0x40001c0c );
 strcasecmp = 0x40007b38;
 strcasestr = 0x40007b7c;
 tolower = 0x40008158;
--- a/components/esp_rom/esp32s3/include/esp32s3/rom/uart.h
+++ b/components/esp_rom/esp32s3/include/esp32s3/rom/uart.h
@@ -111,6 +111,8 @@
     XON_XOFF_CTRL
 } UartFlowCtrl;
 
+/* Workaround: EMPTY conflicts with Zephyr's util_macro.h */
+#undef EMPTY
 typedef enum {
     EMPTY,
     UNDER_WRITE,
--- a/components/esp_rom/esp32s3/ld/esp32s3.rom.newlib.ld
+++ b/components/esp_rom/esp32s3/ld/esp32s3.rom.newlib.ld
@@ -15,14 +15,20 @@
  * multiple files to make it compatible with picolibc.
  */
 
-strdup = 0x40001380;
-strndup = 0x400013ec;
-rand = 0x400014a0;
-srand = 0x400014ac;
-atoi = 0x400014d0;
-atol = 0x400014dc;
-strtol = 0x400014e8;
-strtoul = 0x400014f4;
+/* ZEPHYR: Keep PROVIDE for these symbols: */
+PROVIDE ( atoi = 0x400014d0 );
+PROVIDE ( atol = 0x400014dc );
+PROVIDE ( strdup = 0x40001380 );
+PROVIDE ( strndup = 0x400013ec );
+PROVIDE ( rand_r = 0x40001494 );
+PROVIDE ( rand = 0x400014a0 );
+PROVIDE ( srand = 0x400014ac );
+PROVIDE ( strtol = 0x400014e8 );
+PROVIDE ( strtoul = 0x400014f4 );
+PROVIDE ( longjmp = 0x40001440 );
+PROVIDE ( setjmp = 0x4000144c );
+/*******************************************/
+
 PROVIDE( fflush = 0x40001500 );
 PROVIDE( _fflush_r = 0x4000150c );
 PROVIDE( _fwalk = 0x40001518 );
--- a/components/esp_rom/patches/esp_rom_print.c
+++ b/components/esp_rom/patches/esp_rom_print.c
@@ -8,7 +8,7 @@
 #include <stdbool.h>
 #include <stdarg.h>
 #include <stddef.h>
-#include <sys/param.h>
+#include <zephyr/sys/util.h>
 #include "esp_rom_caps.h"
 #include "esp_rom_sys.h"
 #include "rom/ets_sys.h"
--- a/components/esp_rom/patches/esp_rom_sys.c
+++ b/components/esp_rom/patches/esp_rom_sys.c
@@ -146,7 +146,7 @@
     }
 }
 
-void __attribute__((constructor)) ets_ops_set_rom_patches(void)
+void ets_ops_set_rom_patches(void)
 {
     /* Copy ROM default function table into our patch table */
     memcpy(&ets_ops_patch_table_ptr, ets_ops_table_ptr, sizeof(struct ets_ops));
--- a/components/esp_rom/patches/esp_rom_tlsf.c
+++ b/components/esp_rom/patches/esp_rom_tlsf.c
@@ -163,7 +163,7 @@
  * This function must not be defined as static, as it is marked as "undefined" in the linker flags
  * (to force the linker to integrate the functions of this file inside the final binary)
  */
-void __attribute__((constructor)) tlsf_set_rom_patches(void)
+void tlsf_set_rom_patches(void)
 {
     /* Copy the ROM default table inside our table */
     memcpy(&heap_tlsf_patch_table_ptr, heap_tlsf_table_ptr, sizeof(struct heap_tlsf_stub_table_t));
