--- a/components/esp_hal_ledc/ledc_hal.c
+++ b/components/esp_hal_ledc/ledc_hal.c
@@ -43,7 +43,7 @@
         if (speed_mode == LEDC_LOW_SPEED_MODE) {
             /* If the source clock used by LEDC hardware is not REF_TICK, it is
             * necessary to retrieve the global clock source used. */
-            ledc_slow_clk_sel_t slow_clk;
+            ledc_slow_clk_sel_t slow_clk = LEDC_SLOW_CLK_RC_FAST;
             ledc_hal_get_slow_clk_sel(hal, &slow_clk);
             driver_clk = (ledc_clk_cfg_t)slow_clk;
         }
