--- a/components/ieee802154/driver/esp_ieee802154_dev.c
+++ b/components/ieee802154/driver/esp_ieee802154_dev.c
@@ -6,7 +6,8 @@
 
 #include <string.h>
 #include "sdkconfig.h"
-#include "freertos/portmacro.h"
+#include <zephyr/irq.h>
+#include "esp_intr_alloc.h"
 #include "soc/periph_defs.h"
 #include "soc/soc.h"
 #include "hal/ieee802154_periph.h"
@@ -70,7 +71,8 @@
 static uint8_t s_rx_index = 0;
 static uint8_t s_enh_ack_frame[128];
 static uint8_t s_recent_rx_frame_info_index;
-static portMUX_TYPE s_ieee802154_spinlock = portMUX_INITIALIZER_UNLOCKED;
+static unsigned int s_ieee802154_spinlock;
+static volatile int s_ieee802154_critical_nesting;
 static intr_handle_t s_ieee802154_isr_handle = NULL;
 
 static esp_err_t ieee802154_sleep_init(void);
@@ -734,12 +736,16 @@
 
 IRAM_ATTR void ieee802154_enter_critical(void)
 {
-    portENTER_CRITICAL(&s_ieee802154_spinlock);
+    if (s_ieee802154_critical_nesting++ == 0) {
+        s_ieee802154_spinlock = irq_lock();
+    }
 }
 
 IRAM_ATTR void ieee802154_exit_critical(void)
 {
-    portEXIT_CRITICAL(&s_ieee802154_spinlock);
+    if (--s_ieee802154_critical_nesting == 0) {
+        irq_unlock(s_ieee802154_spinlock);
+    }
 }
 
 IEEE802154_NOINLINE static void ieee802154_isr(void *arg)
