--- a/components/wpa_supplicant/esp_supplicant/src/esp_common.c
+++ b/components/wpa_supplicant/esp_supplicant/src/esp_common.c
@@ -24,7 +24,6 @@
 #include "rsn_supp/wpa.h"
 #include "esp_private/wifi.h"
 #include "esp_wifi_types_generic.h"
-#include "esp_roaming.h"
 
 /* Utility Functions */
 esp_err_t esp_supplicant_str_to_mac(const char *str, uint8_t dest[6])
--- a/components/wpa_supplicant/esp_supplicant/src/esp_dpp.c
+++ b/components/wpa_supplicant/esp_supplicant/src/esp_dpp.c
@@ -771,14 +771,12 @@
     esp_dpp_auth_resp_retry_timeout(NULL, NULL);
 }
 
-static void tx_status_eloop_handler(void *eloop_ctx, void *event_data)
+static void tx_status_handler(void *arg, esp_event_base_t event_base,
+                              int32_t event_id, void *event_data)
 {
     struct dpp_authentication *auth = s_dpp_ctx.dpp_auth;
-    wifi_event_action_tx_status_t *evt = event_data;
 
-    if (!evt) {
-        return;
-    }
+    wifi_event_action_tx_status_t *evt = event_data;
 
     wpa_printf(MSG_DEBUG, "Mgmt Tx Status - %d, Cookie - 0x%x",
                evt->status, (uint32_t)evt->context);
@@ -786,9 +784,13 @@
     if (evt->op_id != s_current_tx_op_id) {
         wpa_printf(MSG_DEBUG, "DPP: status not for recent frame op_id=%u, s_current_tx_op_id=%u",
                    evt->op_id, s_current_tx_op_id);
-    } else if (!auth) {
+        return;
+    }
+    if (!auth) {
         wpa_printf(MSG_DEBUG, "Auth already deinitialized, return");
-    } else if (auth->waiting_auth_conf) {
+        return;
+    }
+    if (auth->waiting_auth_conf) {
         eloop_cancel_timeout(esp_dpp_auth_resp_retry_timeout, NULL, NULL);
         if (evt->status == WIFI_ACTION_TX_FAILED) {
             /* failed to send auth response frame */
@@ -808,53 +810,22 @@
             eloop_register_timeout(ESP_GAS_TIMEOUT_SECS, 0, gas_query_timeout, NULL, auth);
         }
     }
-    os_free(evt);
 }
 
-static void tx_status_handler(void *arg, esp_event_base_t event_base,
-                              int32_t event_id, void *event_data)
-{
-    wifi_event_action_tx_status_t *evt_c = os_malloc(sizeof(*evt_c));
-    if (evt_c) {
-        os_memcpy(evt_c, event_data, sizeof(*evt_c));
-        if (eloop_register_timeout(0, 0, tx_status_eloop_handler, NULL, evt_c) < 0) {
-            os_free(evt_c);
-        }
-    } else {
-        wpa_printf(MSG_ERROR, "DPP: Failed to allocate memory for TX status");
-    }
-}
-
-static void roc_status_eloop_handler(void *eloop_ctx, void *event_data)
+static void roc_status_handler(void *arg, esp_event_base_t event_base,
+                               int32_t event_id, void *event_data)
 {
     wifi_event_roc_done_t *evt = (wifi_event_roc_done_t *)event_data;
 
-    if (evt) {
-        if (evt->context == (uint32_t)s_action_rx_cb) {
-            eloop_cancel_timeout(dpp_listen_next_channel, NULL, NULL);
-            eloop_register_timeout(0, 0, dpp_listen_next_channel, NULL, NULL);
-        }
-        os_free(evt);
+    if (evt->context == (uint32_t)s_action_rx_cb) {
+        eloop_cancel_timeout(dpp_listen_next_channel, NULL, NULL);
+        eloop_register_timeout(0, 0, dpp_listen_next_channel, NULL, NULL);
     }
 
     atomic_store(&roc_in_progress, false);
     os_event_group_set_bits(s_dpp_event_group, DPP_ROC_EVENT_HANDLED);
 }
 
-static void roc_status_handler(void *arg, esp_event_base_t event_base,
-                               int32_t event_id, void *event_data)
-{
-    wifi_event_roc_done_t *evt_c = os_malloc(sizeof(*evt_c));
-    if (evt_c) {
-        os_memcpy(evt_c, event_data, sizeof(*evt_c));
-        if (eloop_register_timeout(0, 0, roc_status_eloop_handler, NULL, evt_c) < 0) {
-            os_free(evt_c);
-        }
-    } else {
-        wpa_printf(MSG_ERROR, "DPP: Failed to allocate memory for ROC status");
-    }
-}
-
 static char *esp_dpp_parse_chan_list(const char *chan_list)
 {
     struct dpp_bootstrap_params_t *params = &s_dpp_ctx.bootstrap_params;
@@ -1124,7 +1095,7 @@
     esp_err_t ret = ESP_OK;
 
     wifi_mode_t mode = 0;
-    if (esp_wifi_get_mode(&mode) || ((mode != WIFI_MODE_STA) && (mode != WIFI_MODE_APSTA))) {
+    if (esp_wifi_get_mode(&mode) || ((mode != ESP32_WIFI_MODE_STA) && (mode != ESP32_WIFI_MODE_APSTA))) {
         wpa_printf(MSG_ERROR, "DPP: failed to init as not in station mode.");
         return ESP_FAIL;
     }
--- a/components/wpa_supplicant/esp_supplicant/src/esp_hostpad_wps.c
+++ b/components/wpa_supplicant/esp_supplicant/src/esp_hostpad_wps.c
@@ -144,7 +144,7 @@
 static int wifi_ap_wps_enable_internal(const esp_wps_config_t *config)
 {
     struct wps_sm *sm = gWpsSm;
-    wifi_mode_t mode = WIFI_MODE_NULL;
+    wifi_mode_t mode = ESP32_WIFI_MODE_NULL;
 
     if (esp_wifi_get_user_init_flag_internal() == 0) {
         wpa_printf(MSG_ERROR, "wps enable: wifi not started cannot enable wpsreg");
@@ -156,7 +156,7 @@
         return ESP_FAIL;
     }
 
-    if (mode != WIFI_MODE_AP && mode != WIFI_MODE_APSTA) {
+    if (mode != ESP32_WIFI_MODE_AP && mode != ESP32_ESP32_WIFI_MODE_APSTA) {
         wpa_printf(MSG_ERROR, "wps enable: mode=%d does not include AP", mode);
         return ESP_ERR_WIFI_MODE;
     }
@@ -265,10 +265,10 @@
 
 static int wifi_ap_wps_start_internal(const unsigned char *pin)
 {
-    wifi_mode_t mode = WIFI_MODE_NULL;
+    wifi_mode_t mode = ESP32_WIFI_MODE_NULL;
 
     esp_wifi_get_mode(&mode);
-    if (mode != WIFI_MODE_AP && mode != WIFI_MODE_APSTA) {
+    if (mode != ESP32_WIFI_MODE_AP && mode != ESP32_ESP32_WIFI_MODE_APSTA) {
         wpa_printf(MSG_ERROR, "wps start: mode=%d does not include AP", mode);
         return ESP_ERR_WIFI_MODE;
     }
--- a/components/wpa_supplicant/esp_supplicant/src/esp_scan.c
+++ b/components/wpa_supplicant/esp_supplicant/src/esp_scan.c
@@ -222,9 +222,9 @@
         } else
 
             if (scan_params->mode == BEACON_REPORT_MODE_PASSIVE) {
-                params->scan_type = WIFI_SCAN_TYPE_PASSIVE;
+                params->scan_type = ESP32_WIFI_SCAN_TYPE_PASSIVE;
             } else {
-                params->scan_type = WIFI_SCAN_TYPE_ACTIVE;
+                params->scan_type = ESP32_WIFI_SCAN_TYPE_ACTIVE;
             }
 
         if (scan_params->bssid) {
--- a/components/wpa_supplicant/esp_supplicant/src/esp_wpa_main.c
+++ b/components/wpa_supplicant/esp_supplicant/src/esp_wpa_main.c
@@ -31,7 +31,9 @@
 #include "esp_eap_client.h"
 #include "esp_common_i.h"
 #include "esp_owe_i.h"
+#ifndef __ZEPHYR__
 #include "esp_roaming.h"
+#endif
 
 #include "esp_wps.h"
 #include "esp_wps_i.h"
--- a/components/wpa_supplicant/esp_supplicant/src/esp_wps.c
+++ b/components/wpa_supplicant/esp_supplicant/src/esp_wps.c
@@ -1,3 +1,4 @@
+
 /*
  * SPDX-FileCopyrightText: 2019-2025 Espressif Systems (Shanghai) CO LTD
  *
@@ -60,7 +61,7 @@
 // Callback structure for handling WPS events within the Wi-Fi task
 static struct wps_sm_funcs *s_wps_sm_cb = NULL;
 
-const char *wps_model_number = CONFIG_IDF_TARGET;
+const char *wps_model_number = CONFIG_SOC_SERIES;
 
 /* API lock to ensure thread safety for public WPS functions */
 void *s_wps_api_lock = NULL;  /* Used in WPS/WPS-REG public API only, never be freed */
@@ -255,9 +256,9 @@
     }
 
     esp_wifi_get_mode(&op_mode);
-    if ((op_mode != WIFI_MODE_STA)
+    if ((op_mode != ESP32_WIFI_MODE_STA)
 #ifdef CONFIG_ESP_WIFI_SOFTAP_SUPPORT
-            && (op_mode != WIFI_MODE_APSTA)
+            && (op_mode != ESP32_WIFI_MODE_APSTA)
 #endif
        ) {
         return false;
@@ -954,9 +955,9 @@
         }
     }
 
-    os_snprintf(s_factory_info->manufacturer, WPS_MAX_MANUFACTURER_LEN, "ESPRESSIF");
-    os_snprintf(s_factory_info->model_name, WPS_MAX_MODEL_NUMBER_LEN, "ESPRESSIF IOT");
-    os_snprintf(s_factory_info->model_number, WPS_MAX_MODEL_NAME_LEN, wps_model_number);
+    os_snprintf(s_factory_info->manufacturer, WPS_MAX_MANUFACTURER_LEN, "%s", "ESPRESSIF");
+    os_snprintf(s_factory_info->model_name, WPS_MAX_MODEL_NUMBER_LEN, "%s", "ESPRESSIF IOT");
+    os_snprintf(s_factory_info->model_number, WPS_MAX_MODEL_NAME_LEN, "%s", wps_model_number);
     os_snprintf(s_factory_info->device_name, WPS_MAX_DEVICE_NAME_LEN, "%s STATION", wps_model_number);
 
     return ESP_OK;
@@ -1033,28 +1034,28 @@
         ret = ESP_FAIL;
         goto _out;
     }
-    os_snprintf(dev->manufacturer, WPS_MAX_MANUFACTURER_LEN, s_factory_info->manufacturer);
+    os_snprintf(dev->manufacturer, WPS_MAX_MANUFACTURER_LEN, "%s", s_factory_info->manufacturer);
 
     dev->model_name = os_zalloc(WPS_MAX_MODEL_NAME_LEN);
     if (!dev->model_name) {
         ret = ESP_FAIL;
         goto _out;
     }
-    os_snprintf(dev->model_name, WPS_MAX_MODEL_NAME_LEN, s_factory_info->model_name);
+    os_snprintf(dev->model_name, WPS_MAX_MODEL_NAME_LEN, "%s", s_factory_info->model_name);
 
     dev->model_number = os_zalloc(WPS_MAX_MODEL_NAME_LEN);
     if (!dev->model_number) {
         ret = ESP_FAIL;
         goto _out;
     }
-    os_snprintf(dev->model_number, WPS_MAX_MODEL_NAME_LEN, s_factory_info->model_number);
+    os_snprintf(dev->model_number, WPS_MAX_MODEL_NAME_LEN, "%s", s_factory_info->model_number);
 
     dev->device_name = os_zalloc(WPS_MAX_DEVICE_NAME_LEN);
     if (!dev->device_name) {
         ret = ESP_FAIL;
         goto _out;
     }
-    os_snprintf(dev->device_name, WPS_MAX_DEVICE_NAME_LEN, s_factory_info->device_name);
+    os_snprintf(dev->device_name, WPS_MAX_DEVICE_NAME_LEN, "%s", s_factory_info->device_name);
 
     dev->serial_number = os_zalloc(16);
     if (!dev->serial_number) {
@@ -1531,9 +1532,9 @@
 
     if (
 #ifdef CONFIG_ESP_WIFI_SOFTAP_SUPPORT
-        mode == WIFI_MODE_AP ||
+        mode == ESP32_WIFI_MODE_AP ||
 #endif
-        mode == WIFI_MODE_NULL) {
+        mode == ESP32_WIFI_MODE_NULL) {
         wpa_printf(MSG_ERROR, "WPS: Invalid Wi-Fi mode for WPS: %d ", mode);
         return ESP_ERR_WIFI_MODE;
     }
--- a/components/wpa_supplicant/port/eloop.c
+++ b/components/wpa_supplicant/port/eloop.c
@@ -12,6 +12,7 @@
  */
 
 #include "includes.h"
+#include <zephyr/kernel.h>
 
 #include "common.h"
 #include "list.h"
@@ -507,7 +508,7 @@
     atomic_store(&eloop.eloop_started, false);
 
     while (atomic_load(&eloop.timeout_running)) {
-        vTaskDelay(100 / portTICK_PERIOD_MS);  // Yield CPU
+        k_msleep(100);
     }
     dl_list_for_each_safe(timeout, prev, &eloop.timeout,
                           struct eloop_timeout, list) {
--- a/components/wpa_supplicant/port/include/os.h
+++ b/components/wpa_supplicant/port/include/os.h
@@ -17,12 +17,22 @@
 #include <sys/types.h>
 #include "esp_types.h"
 #include <string.h>
+#include <strings.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include "esp_err.h"
 #include "supplicant_opt.h"
 #include "esp_private/esp_wifi_private.h"
 #include "esp_wifi.h"
+#include "esp_heap_adapter.h"
+
+#ifdef __ZEPHYR__
+/* With -std=c17 these POSIX/BSD functions are not declared by default */
+char *strdup(const char *);
+size_t strlcpy(char *, const char *, size_t);
+
+char *z_strdup(const char *s);
+#endif
 
 /* Modifying datatype for platform and compiler independence */
 
@@ -213,31 +223,31 @@
  */
 
 #ifndef os_malloc
-#define os_malloc(s) malloc((s))
+#define os_malloc(s) os_wpa_malloc_func((s))
 #endif
 #ifndef os_realloc
-#define os_realloc(p, s) realloc((p), (s))
+#define os_realloc(p, s) os_wpa_realloc_func((p), (s))
 #endif
 #ifndef os_zalloc
-#define os_zalloc(s) calloc(1, (s))
+#define os_zalloc(s) os_wpa_calloc_func(1, (s))
 #endif
 #ifndef os_calloc
-#define os_calloc(p, s) calloc((p), (s))
+#define os_calloc(p, s) os_wpa_calloc_func((p), (s))
 #endif
 
 #ifndef os_free
-#define os_free(p) free((p))
+#define os_free(p) os_wpa_free_func((p))
 #endif
 
 #ifndef os_bzero
-#define os_bzero(s, n) bzero(s, n)
+#define os_bzero(s, n) memset(s, 0, n)
 #endif
 
 #ifndef os_strdup
 #ifdef _MSC_VER
 #define os_strdup(s) _strdup(s)
 #else
-#define os_strdup(s) strdup(s)
+#define os_strdup(s) z_strdup(s)
 #endif
 #endif
 char * ets_strdup(const char *s);
--- a/components/wpa_supplicant/port/os_xtensa.c
+++ b/components/wpa_supplicant/port/os_xtensa.c
@@ -24,40 +24,53 @@
 
 #include "os.h"
 #include <stdlib.h>
-#include <unistd.h>
-#include <sys/time.h>
-#include "esp_random.h"
+#include "esp_system.h"
 #include "utils/common.h"
+#include <zephyr/random/random.h>
 #include "mbedtls/platform_util.h"
 
 int os_get_time(struct os_time *t)
 {
-    struct timeval tv;
-    int ret = gettimeofday(&tv, NULL);
-    t->sec = (os_time_t) tv.tv_sec;
-    t->usec = tv.tv_usec;
-    return ret;
+	if (t == NULL) {
+		return -1;
+	}
+
+	int64_t now = k_uptime_ticks();
+	t->sec = now / CONFIG_SYS_CLOCK_TICKS_PER_SEC;
+	t->usec = k_ticks_to_us_floor64(now);
+
+	return 0;
 }
 
 unsigned long os_random(void)
 {
-    return esp_random();
+	return sys_rand32_get();
 }
 
 int os_get_random(unsigned char *buf, size_t len)
 {
-    esp_fill_random(buf, len);
-    return 0;
+	sys_rand_get((void *)buf, len);
+	return 0;
 }
 
 void os_sleep(os_time_t sec, os_time_t usec)
 {
     if (sec) {
-        sleep(sec);
+        k_sleep(K_SECONDS(sec));
     }
     if (usec) {
-        usleep(usec);
+        k_sleep(K_USEC(usec));
+    }
+}
+
+char *z_strdup(const char *s)
+{
+    size_t size = strlen(s) + 1;
+    char *p = k_malloc(size);
+    if (p) {
+        memcpy(p, s, size);
     }
+    return p;
 }
 
 #ifdef CONFIG_CRYPTO_MBEDTLS
--- a/components/wpa_supplicant/src/utils/common.h
+++ b/components/wpa_supplicant/src/utils/common.h
@@ -456,7 +456,9 @@
 	unsigned int num;
 };
 
+#ifndef ARRAY_SIZE
 #define ARRAY_SIZE(a) (sizeof(a) / sizeof((a)[0]))
+#endif
 #ifndef TEST_FAIL
 #define TEST_FAIL() 0
 #endif
